// @ihf-rivendell/qa/dist/integration/api/requests/ContractsRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContractsRequests = void 0;
const BaseRequest_1 = require("./BaseRequest");
var ProductType;
(function (ProductType) {
    ProductType["Loan"] = "LOAN";
})(ProductType || (ProductType = {}));
var ContractVersion;
(function (ContractVersion) {
    ContractVersion["v1"] = "v1";
})(ContractVersion || (ContractVersion = {}));
class ContractsRequests extends BaseRequest_1.BaseRequests {
    props;
    contractsRepository;
    customerId;
    loanId;
    constructor(props, shouldPrintRequest) {
        super(shouldPrintRequest);
        this.props = props;
        this.contractsRepository = props.repositories.repository;
        this.customerId = props.source.customer.id;
        this.loanId = props.source.loan.id;
    }
    async createContract() {
        const buildRequest = this.buildCreateContractPromise();
        await this.printRequest(buildRequest);
        return this.contractsRepository.createContract(buildRequest);
    }
    async contractSignature(contractId) {
        const buildRequest = this.buildContractSignature(contractId);
        await this.printRequest(buildRequest);
        return this.contractsRepository.updateStatusSigned(buildRequest);
    }
    async findContract(contractId) {
        return this.contractsRepository.findById(this.customerId, contractId);
    }
    buildContractSignature(contractId) {
        return {
            id: contractId,
            customer: {
                id: this.customerId,
            },
            product: {
                type: ProductType.Loan,
                id: this.loanId,
            },
            version: ContractVersion.v1,
        };
    }
    buildCreateContractPromise() {
        return {
            customer: {
                id: this.customerId,
            },
            product: {
                type: ProductType.Loan,
                id: this.loanId,
            },
            version: ContractVersion.v1,
        };
    }
}
exports.ContractsRequests = ContractsRequests;
