// @ihf-rivendell/qa/dist/repositories/CognitoRepository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CognitoRepository = void 0;
const client_cognito_identity_provider_1 = require("@aws-sdk/client-cognito-identity-provider");
const BaseRepository_1 = require("./BaseRepository");
class CognitoRepository extends BaseRepository_1.BaseRepository {
    props;
    constructor(props) {
        super(props);
        this.props = props;
    }
    async changePassword(username, password) {
        await this.props.cognitoClient.send(new client_cognito_identity_provider_1.AdminSetUserPasswordCommand({
            UserPoolId: this.props.userPoolId,
            Username: username,
            Password: password,
            Permanent: true,
        }));
    }
    async initiateAuth(number, password) {
        const result = await this.props.cognitoClient.send(new client_cognito_identity_provider_1.InitiateAuthCommand({
            ClientId: this.props.applications.internetBanking,
            AuthFlow: "USER_PASSWORD_AUTH",
            AuthParameters: {
                USERNAME: number,
                PASSWORD: password,
            },
        }));
        const token = result.AuthenticationResult?.IdToken;
        return { token };
    }
}
exports.CognitoRepository = CognitoRepository;
