// @ihf-rivendell/qa/dist/repositories/EventsRepository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventsRepository = void 0;
const util_utf8_1 = require("@aws-sdk/util-utf8");
const client_lambda_1 = require("@aws-sdk/client-lambda");
const httpConstants_1 = require("../integration/assets/httpConstants");
const BaseRepository_1 = require("./BaseRepository");
class EventsRepository extends BaseRepository_1.BaseRepository {
    props;
    constructor(props) {
        super(props);
        this.props = props;
    }
    async find(correlationId, label) {
        const response = await this.props.lambdaClient.send(new client_lambda_1.InvokeCommand({
            FunctionName: this.props.eventsLambda,
            Payload: (0, util_utf8_1.fromUtf8)(JSON.stringify({
                resource: "/events",
                httpMethod: httpConstants_1.HttpMethod.GET,
                queryStringParameters: { correlationId, label },
            })),
        }));
        return this.processLambdaApiResponse(response);
    }
}
exports.EventsRepository = EventsRepository;
